gusto/core/io.py

```diff
@@ -917,3 +922,87 @@ def topo_sort(field_deps):
                           if f not in schedule)
         raise RuntimeError("Field dependencies have a cycle:\n\n%s" % cycle)
     return list(map(name2field.__getitem__, schedule))
+
+
+def make_nc_dataset(filename, access, comm):
+    """Create a netCDF data set, possibly in parallel.
+
+    Args:
+        filename (str): The filename.
+        access (str): The access descriptor - ``r``, ``w`` or ``a``.
+        comm: The communicator.
+
+    Returns:
+        tuple: 2-tuple of :class:`netCDF4_netCDF4.Dataset` (or `None`) and `bool`
+            indicating whether netCDF supports parallel usage. If parallel is not
+            supported then the dataset will be `None` on all but rank 0.
+
+    A warning will be thrown if this function is called in parallel and a
+    non-parallel netCDF4 is installed.
+
+    """
+    try:
+        nc_field_file = Dataset(filename, access, parallel=True)
+        nc_supports_parallel = True
+    except ValueError:
+        # parallel netCDF not available, use the serial version instead
+        if comm.size > 1:
+            import warnings
+            warnings.warn(
+                "Serial netCDF in use even though you are running in parallel. This "
+                "is especially inefficient at high core counts. Please refer to the "
+                "documentation for information about installing a parallel version "
+                "of netCDF.",
+                UserWarning,
+            )
+
+        if comm.rank == 0:
+            nc_field_file = Dataset(filename, access, parallel=False)
+        else:
+            nc_field_file = None
+        nc_supports_parallel = False
+    return nc_field_file, nc_supports_parallel
+
+
+def gather_field_data(field, field_index, domain):
+    """Gather global field data into a single array on rank 0.
+
+    Args:
+        field (:class:`firedrake.Function`): The field to gather.
+        field_index (int): Index used to identify the field.
+        domain (:class:`Domain`): The domain.
+
+    Returns:
+        :class:`numpy.ndarray` that is the concatenation of all DoFs on
+        all ranks.
+
+    Note that this operation is *extremely inefficient* when run with large
+    amounts of parallelism. Avoid calling if at all possible.
+
+    """
+    comm = domain.mesh.comm
+
+    if comm.size == 1:
+        return field.dat.data_ro
+
+    space_name = field.function_space().name
+
+    if comm.rank == 0:
```

Comment:
This could also be done with `comm.Allgatherv` to handle the different processors - this will also allow the MPI to pick how to do the gather.

Reply:
We only want to gather, not allgather, but that would be better.

Reply:
Oh, yeah that's what I meant, `Gatherv`. It's by no means essential but it's a bit cleaner - less code is easier to look after!

Reply:
I looked into `Gatherv` but the interface appears very confusing and we still need to think about bounds. If we use `comm.gather` followed by `np.concatenate` it seems to work very nicely.

Reply:
Much nicer.